R/telemetry.R

```diff
@@ -0,0 +1,587 @@
+    #' Shiny session.
+
+    log_button = function(
+      input_id, session = shiny::getDefaultReactiveDomain()
+    ) {
+      self$log_input(input_id, session = session)
+    },
+
+    #' @description
+    #' A short description...
+    #'
+    #' @param track_values flag that indicates if the basic telemetry should
+    #' track the values of the inputs that are changing. `FALSE` by default.
+    #' This parameter is ignored if `track_inputs` is `FALSE`.
+    #' @param excluded_inputs vector of input_ids that should not be tracked.
+    #' By default it doesn't track browser version, which is added by this
+    #' package.
+    #' @param session ShinySession object or NULL to identify the current
+    #' Shiny session.
+
+    log_all_inputs = function(
+      track_values = FALSE,
+      excluded_inputs = c("browser_version"),
+      session = shiny::getDefaultReactiveDomain()
+    ) {
+      input <- shiny::reactiveValues()
+      if (checkmate::test_r6(session, "ShinySession")) {
+        input <- session$input
+      }
+
+      input_values <- shiny::isolate(shiny::reactiveValuesToList(input))
+
+      if (checkmate::test_r6(session, "ShinySession")) {
+        session$userData$shiny_input_values <- input_values
+      }
+
+      logger::log_debug(logger::skip_formatter(
+        paste(
+          "Default Shiny inputs initialized:",
+          as.character(jsonlite::toJSON(input_values, auto_unbox = TRUE))
+        )),
+        namespace = "shiny.telemetry"
+      )
+
+      shiny::observe({
+        if (checkmate::test_r6(session, "ShinySession")) {
+          old_input_values <- session$userData$shiny_input_values
+        } else {
+          old_input_values <- input_values
+        }
+        new_input_values <- shiny::reactiveValuesToList(input)
+
+        if (NROW(new_input_values) != 0) {
+          names <- unique(c(names(old_input_values), names(new_input_values)))
+          names <- setdiff(names, excluded_inputs)
+          for (name in names) {
+            old <- old_input_values[name]
+            new <- new_input_values[name]
+            if (!identical(old, new)) {
+              if (isTRUE(track_values)) {
+                logger::log_debug(
+                  "Shiny input change detected on {name} ",
+                  "{old[[name]]} -> {new[[name]]}",
+                  namespace = "shiny.telemetry"
+                )
+              } else {
+                logger::log_debug(
+                  "Shiny input change detected on {name} (no value tracking) ",
+                  "{old[[name]]} -> {new[[name]]}",
+                  namespace = "shiny.telemetry"
+                )
+              }
+
+              if (isTRUE(track_values)) {
+                private$.log_event(
+                  event = "input",
+                  id = name,
+                  value = new[[name]],
+                  session = session
+                )
+                next
```

Comment:
Wouldn't it be more readable to just put the code after this if in `else`? :wink:

Reply:
refactored

```R
if (isTRUE(track_values)) {
                logger::log_debug(
                  "event: input change: {name} ",
                  ":: {old[[name]]} -> {new[[name]]}",
                  namespace = "shiny.telemetry"
                )
                log_value <- new[[name]]
              } else {
                logger::log_debug(
                  "event: input change: {name} :: (no value tracking)",
                  namespace = "shiny.telemetry"
                )
              }

              private$.log_event(
                event = "input",
                id = name,
                value = log_value,
                session = session
              )
```

Reply:
refactored

```R
if (isTRUE(track_values)) {
                logger::log_debug(
                  "event: input change: {name} ",
                  ":: {old[[name]]} -> {new[[name]]}",
                  namespace = "shiny.telemetry"
                )
                log_value <- new[[name]]
              } else {
                logger::log_debug(
                  "event: input change: {name} :: (no value tracking)",
                  namespace = "shiny.telemetry"
                )
              }

              private$.log_event(
                event = "input",
                id = name,
                value = log_value,
                session = session
              )
```

Reply:
refactored

```R
if (isTRUE(track_values)) {
                logger::log_debug(
                  "event: input change: {name} ",
                  ":: {old[[name]]} -> {new[[name]]}",
                  namespace = "shiny.telemetry"
                )
                log_value <- new[[name]]
              } else {
                logger::log_debug(
                  "event: input change: {name} :: (no value tracking)",
                  namespace = "shiny.telemetry"
                )
              }

              private$.log_event(
                event = "input",
                id = name,
                value = log_value,
                session = session
              )
```